File: Day_1/main.cpp
// Given a list of numbers and a number k, 
// return whether any two numbers from the list add up to k.

// For example, given [10, 15, 3, 7] and k of 17, 
// return true since 10 + 7 is 17.

// Bonus: Can you do this in one pass?

#include <iostream>

bool is_two_num_equal(int num, int *list, int size)
{
    for (size_t i = 0; i < 5; i++)
    {
        for (size_t j = 0; j < 5; j++)
        {
            if (i == j)
            {
                continue;
            }
            if (list[i] + list[j] == num)
            {
                return true;
            }
        }
    }
    return false;
}

int main(int argc, char const *argv[])
{
    int num = 5;
    int list[] = {1,2,10,1,3};
    int size = sizeof(list)/sizeof(list[0]);
    std::cout << is_two_num_equal(num, list, size) << std::endl;
    return 0;
}

// Explanation

// Function `is_two_num_equal` takes in three parameters, num, list, and size
// This will loop over list (using size as a limiter) in a nested loop, skipping
// over the number if it shares an index, and while add two numbers together
// returning true is they equal num, and false if all options are exhausted

// Satisfaction: 3/10 - C++ is hard

File: Day_1/main.py
# Given a list of numbers and a number k, 
# return whether any two numbers from the list add up to k.

# For example, given [10, 15, 3, 7] and k of 17, 
# return true since 10 + 7 is 17.

# Bonus: Can you do this in one pass?

def is_two_num_equal(num, list):
    def mapper(n, pos):
        return [n+i for num, i in enumerate(list) if num != pos]
    result = [mapper(i, pos) for pos, i in enumerate(list)]
    return num in [i for s in result for i in s]
    
if __name__ == "__main__":
    print(is_two_num_equal(5, [1,2,1,1,6]))

# Explanation:
# 
# Function `is_two_num_equal` takes in two parameters, num and list
# This function gets `result` with list comprehension by getting the result
# of `mapper`, in which each number from `list` and its position is passed
# It then flattens the now 2D list into a 1D list, and returns if `num` is in it
# 
# A nested function, `mapper` is defined and takes two parameters, n and pos
# It uses list comprehension to generate and return a list composed of each number 
# in `list` added to `n` except for the number that shares its position
# 
# Satisfaction: 6/10 - Honestly it's a mess